apiserver/facades/client/secrets/secrets.go

```diff
@@ -406,16 +407,48 @@ func (s *SecretsAPI) updateSecret(backend provider.SecretsBackend, arg params.Up
 			}
 		}
 	}
-	_, err = s.secretsState.UpdateSecret(uri, fromUpsertParams(arg.UpsertSecretArg))
-	return errors.Trace(err)
+	md, err = s.secretsState.UpdateSecret(uri, fromUpsertParams(arg.AutoPrune, arg.UpsertSecretArg))
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if md.AutoPrune {
+		// If the secret was updated, we need to delete the old unused secret revisions.
+		revsToDelete, err := s.secretsState.ListUnusedSecretRevisions(uri)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		pruneArg := params.DeleteSecretArg{URI: md.URI.String()}
+		for _, rev := range revsToDelete {
+			if rev == md.LatestRevision {
+				// We don't want to delete the latest revision.
+				continue
+			}
+			pruneArg.Revisions = append(pruneArg.Revisions, rev)
+		}
+		if len(pruneArg.Revisions) == 0 {
+			return nil
+		}
+		pruneResult, err := s.RemoveSecrets(params.DeleteSecretArgs{Args: []params.DeleteSecretArg{pruneArg}})
+		if err != nil {
+			// We don't want to fail the update if we can't prune the unused secret revisions because they will be picked up later
+			// when the secret has any new obsolute revisions.
+			logger.Warningf("failed to prune unused secret revisions for %q: %v", uri, err)
+		}
+		if err = pruneResult.Combine(); err != nil {
+			logger.Warningf("failed to prune unused secret revisions for %q: %v", uri, pruneResult.Combine())
+		}
+	}
+	return nil
 }
 
 // RemoveSecrets isn't on the v1 API.
 func (s *SecretsAPIV1) RemoveSecrets(_ struct{}) {}
 
 // RemoveSecrets remove user secret.
 func (s *SecretsAPI) RemoveSecrets(args params.DeleteSecretArgs) (params.ErrorResults, error) {
-	return commonsecrets.RemoveSecretsForClient(
+	// ??? should we allow to delete secrets/revisions if they are still been used???
```

Comment:
I think it's ok so long as artefacts like ACLs etc are removed. For user secrets, we can prompt to confirm. For charm secrets, they can just delete whatever they want at the moment.

Reply:
Created card `JUJU-4719` for tracking this and also for existing unit/app-owned secrets removal.